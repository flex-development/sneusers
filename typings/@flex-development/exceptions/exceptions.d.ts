declare module '@flex-development/exceptions/exceptions' {
  export { ValidationException } from '@flex-development/exceptions/exceptions/index'
}
