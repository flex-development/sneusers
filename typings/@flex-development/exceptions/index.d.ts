/// <reference path='./exceptions.d.ts' />

declare module '@flex-development/exceptions' {
  export { ValidationException } from '@flex-development/exceptions/exceptions/index'
}
