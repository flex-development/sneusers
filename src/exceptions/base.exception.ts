import {
  isExceptionCode,
  isExceptionJSON
} from '@flex-development/exceptions/guards'
import { NullishString, ObjectPlain } from '@flex-development/tutils'
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger'
import { ExceptionDataDTO } from '@sneusers/dtos'
import {
  ExceptionClassName,
  ExceptionCode,
  ExceptionId,
  SequelizeErrorCode
} from '@sneusers/enums'
import { ExceptionJSON } from '@sneusers/interfaces'
import { ExceptionData, ExceptionErrors, SequelizeError } from '@sneusers/types'
import omit from 'lodash.omit'
import type { ValidationErrorItem } from 'sequelize'

/**
 * @file Filters - Exception
 * @module sneusers/filters/Exception
 */

/**
 * Base exception class.
 *
 * @template T - Aggregated error type
 */
export default class Exception<T = any> {
  /**
   * @static
   * @readonly
   * @property {string} DEFAULT_MESSAGE - Default error {@link message}
   */
  static readonly DEFAULT_MESSAGE: string = 'Unknown error'

  /**
   * @readonly
   * @property {ExceptionClassName} className - Associated CSS class name
   */
  @ApiProperty({
    description: 'Associated CSS class name',
    enum: ExceptionClassName,
    enumName: 'ExceptionClassName'
  })
  readonly className: ExceptionClassName

  /**
   * @readonly
   * @property {ExceptionCode} code - HTTP error response status code
   */
  @ApiProperty({
    description: 'HTTP error response status code',
    enum: ExceptionCode,
    enumName: 'ExceptionCode'
  })
  readonly code: ExceptionCode

  /**
   * @readonly
   * @property {ExceptionData} data - Custom error data
   */
  @ApiProperty({ description: 'Custom error data', type: Object })
  readonly data: ExceptionData

  /**
   * @readonly
   * @property {ExceptionErrors<T>} errors - Aggregated errors
   */
  @ApiProperty({ description: 'Aggregated errors', type: [Object] })
  readonly errors: ExceptionErrors<T>

  /**
   * @readonly
   * @property {ExceptionId} id - HTTP error response type
   */
  @ApiProperty({
    description: 'HTTP error response type',
    enum: ExceptionId,
    enumName: 'ExceptionId',
    name: 'name'
  })
  readonly id: ExceptionId

  /**
   * @readonly
   * @property {string} message - Error message
   */
  @ApiProperty({ description: 'Reason exception was thrown', type: String })
  readonly message: string

  /**
   * @readonly
   * @property {'Exception'} name - Constructor name
   */
  readonly name: 'Exception' = 'Exception'

  /**
   * @readonly
   * @property {string} stack - Stack trace
   */
  @ApiPropertyOptional({ description: 'Stack trace', type: String })
  readonly stack?: string

  /**
   * Instantiates a new `Exception`.
   *
   * @param {ExceptionCode} [code=500] - HTTP error response status code
   * @param {NullishString} [message=Exception.DEFAULT_MESSAGE] - Error message
   * @param {ExceptionDataDTO<T> | ObjectPlain} [data={}] - Custom error data
   * @param {ExceptionCode} [data.code] - Override error response status code
   * @param {ExceptionErrors<T>} [data.errors] - Single error or group of errors
   * @param {string} [data.message] - Custom error message. Overrides `message`
   * @param {string} [stack] - Error stack
   */
  constructor(
    code: ExceptionCode = ExceptionCode.INTERNAL_SERVER_ERROR,
    message: NullishString = Exception.DEFAULT_MESSAGE,
    data: ExceptionDataDTO<T> | ObjectPlain = {},
    stack?: string
  ) {
    this.code = data.code || Exception.formatCode(code)
    this.errors = [data.errors || []].flat() as ExceptionErrors<T>
    this.id = Exception.idByCode(this.code)
    this.message = data.message || message || Exception.DEFAULT_MESSAGE
    this.stack = stack
    this.className = ExceptionClassName[this.id]
    this.data = omit(data, ['code', 'errors', 'message'])

    // If data is actually ExceptionJSON, override previously set properties
    if (isExceptionJSON(data)) {
      this.className = data.className
      this.code = data.code
      this.data = data.data
      this.errors = data.errors as ExceptionErrors<T>
      this.id = data.name
      this.message = data.message
    }
  }

  /**
   * Returns {@link ExceptionCode.INTERNAL_SERVER_ERROR} if `code` is not a
   * valid {@link ExceptionCode}.
   *
   * @param {any} [code] - Possible status code
   * @return {ExceptionCode} Exception status code
   */
  static formatCode(code?: any): ExceptionCode {
    return isExceptionCode(code) ? code : ExceptionCode.INTERNAL_SERVER_ERROR
  }

  /**
   * Transforms a {@link SequelizeError}, `error`, into an {@link Exception}.
   *
   * @template T - Sequelize error type
   *
   * @param {SequelizeError} error - Sequelize error
   * @param {ExceptionDataDTO<T> | ObjectPlain} [data={}] - Custom error data
   * @param {ExceptionCode} [data.code] - Override error response status code
   * @param {ExceptionErrors<T>} [data.errors] - Single error or group of errors
   * @param {string} [data.message] - Custom error message. Overrides `message`
   * @return {Exception<T>} New `Exception` instance
   */
  static fromSequelizeError<
    T extends SequelizeError | ValidationErrorItem = SequelizeError
  >(
    error: SequelizeError,
    data: ExceptionDataDTO<T> | ObjectPlain = {}
  ): Exception<T> {
    if (!data.errors && error['errors']) data.errors = error['errors']

    data = { error: error.name, errors: data.errors || [error], ...data }

    return new Exception<T>(
      SequelizeErrorCode[error.name],
      error.message,
      data,
      error.stack
    )
  }

  /**
   * Finds a HTTP error response status type by status code. If a type isn't
   * found, {@link ExceptionId.INTERNAL_SERVER_ERROR} will be returned.
   *
   * @param {ExceptionCode} code - HTTP error response status code
   * @return {ExceptionId} - HTTP exception type
   */
  static idByCode(code: ExceptionCode): ExceptionId {
    const id = ExceptionId.INTERNAL_SERVER_ERROR
    const ids = Object.keys(ExceptionId) as ExceptionId[]

    return ids.find(id => ExceptionCode[id] === code) || id
  }

  /**
   * Returns a JSON object representing the current Exception.
   *
   * To help identify JSON representations of {@link Exception} class objects,
   * the {@link data}  will have an `isExceptionJSON` property added.
   *
   * @return {ExceptionJSON<T>} JSON representation of exception
   */
  toJSON(): ExceptionJSON<T> {
    const json = {} as ExceptionJSON<T>

    Object.assign(json, { name: this.id })
    Object.assign(json, { message: this.message })
    Object.assign(json, { code: this.code })
    Object.assign(json, { className: this.className })
    Object.assign(json, { data: { ...this.data, isExceptionJSON: true } })
    Object.assign(json, { errors: Object.freeze(this.errors) })
    Object.assign(json, { stack: this.stack })

    return json
  }

  /**
   * Returns the exception status {@link code}.
   *
   * @return {ExceptionCode} HTTP error response status code
   */
  get status(): ExceptionCode {
    return this.code
  }
}
