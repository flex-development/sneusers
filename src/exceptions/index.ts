/**
 * @file Entry Point - Exceptions
 * @module sneusers/exceptions
 */

export { default as Exception } from './base.exception'
