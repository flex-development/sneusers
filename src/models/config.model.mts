/**
 * @file Models - Config
 * @module sneusers/models/Config
 */

import bcrypt from 'bcrypt'
import {
  IsIn,
  IsNotEmpty,
  IsString,
  validateSync as validate,
  type ValidationError
} from 'class-validator'

/**
 * Configuration environment.
 *
 * @class
 */
class Config {
  /**
   * The hostname or IP address where the application will listen for incoming
   * connections.
   *
   * @default '127.0.0.1'
   *
   * @public
   * @instance
   * @member {string} HOST
   */
  @IsString()
  @IsNotEmpty()
  public HOST!: string

  /**
   * Name for {@linkcode HOST}.
   *
   * @default 'localhost'
   *
   * @public
   * @instance
   * @member {string} HOSTNAME
   */
  @IsString()
  @IsNotEmpty()
  public HOSTNAME!: string

  /**
   * Access token expiration time (in seconds).
   *
   * @public
   * @instance
   * @member {number} JWT_EXPIRY
   */
  public JWT_EXPIRY: number

  /**
   * Refresh token expiration time (in seconds).
   *
   * @public
   * @instance
   * @member {number} JWT_EXPIRY_REFRESH
   */
  public JWT_EXPIRY_REFRESH: number

  /**
   * Web token signing secret.
   *
   * @public
   * @instance
   * @member {string} JWT_SECRET
   */
  public JWT_SECRET: string

  /**
   * The type of environment the Node.js process is running in.
   *
   * @default 'development'
   *
   * @public
   * @instance
   * @member {string} NODE_ENV
   */
  @IsIn(['development', 'test', 'production'])
  public NODE_ENV!: string

  /**
   * The port number the application will listen for incoming connections on.
   *
   * @default 8080
   *
   * @public
   * @instance
   * @member {number} PORT
   */
  public PORT: number

  /**
   * The url the application will listen for incoming connections on.
   *
   * @public
   * @instance
   * @member {URL} URL
   */
  public URL: URL

  /**
   * Create a new configuration environment.
   *
   * @param {NodeJS.Dict<string | null | undefined>} env
   *  Environment variables object
   * @throws {AggregateError}
   */
  constructor(env: NodeJS.Dict<string | null | undefined>) {
    Object.assign(this, env)

    this.HOST = (this.HOST || '127.0.0.1').trim()
    this.HOSTNAME = (this.HOSTNAME || 'localhost').trim()
    this.NODE_ENV = (this.NODE_ENV || 'development').trim()

    /**
     * Validation errors.
     *
     * @const {ValidationError[]} errors
     */
    const errors: ValidationError[] = validate(this, {
      enableDebugMessages: true,
      forbidUnknownValues: true,
      skipMissingProperties: false,
      stopAtFirstError: false,
      validationError: { target: false },
      whitelist: true
    })

    if (errors.length) {
      throw new AggregateError(errors, 'Invalid app configuration')
    }

    this.JWT_EXPIRY = 900
    this.JWT_EXPIRY_REFRESH = 86400
    this.JWT_SECRET = bcrypt.genSaltSync()
    this.PORT = 8080

    this.URL = new URL('http://' + this.HOSTNAME)
    if (this.HOSTNAME === 'localhost') this.URL.port = this.PORT.toString()
  }
}

export default Config
