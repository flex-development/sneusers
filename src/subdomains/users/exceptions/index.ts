/**
 * @file Users Subdomain Entry Point - Exceptions
 * @module sneusers/subdomains/users/exceptions
 */

export { default as UniqueEmailException } from './unique-email.exception'
