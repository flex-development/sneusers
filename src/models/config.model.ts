/**
 * @file Data Models - Config
 * @module sneusers/models/Config
 */

import type { IConfig } from '#src/interfaces'
import { Exception } from '@flex-development/exceptions'
import { AppEnv, NodeEnv, isUndefined } from '@flex-development/tutils'
import {
  IsEnum,
  IsNotEmpty,
  IsNumber,
  IsString,
  IsUrl,
  validateSync,
  type ValidationError
} from 'class-validator'
import { shake } from 'radash'

/**
 * Data model representing environment variables used by the application.
 *
 * @class
 * @implements {IConfig}
 */
class Config implements IConfig {
  /**
   * Application environment.
   *
   * @see {@linkcode AppEnv}
   *
   * @default AppEnv.DEV
   *
   * @public
   * @instance
   * @member {AppEnv} APP_ENV
   */
  @IsEnum(AppEnv)
  public APP_ENV: AppEnv

  /**
   * Database hostname.
   *
   * @default 'mongo'
   *
   * @public
   * @instance
   * @member {string} DB_HOSTNAME
   */
  @IsString()
  @IsNotEmpty()
  public DB_HOSTNAME: string

  /**
   * Database name.
   *
   * @default APP_ENV
   *
   * @public
   * @instance
   * @member {string} DB_NAME
   */
  @IsString()
  @IsNotEmpty()
  public DB_NAME: string

  /**
   * Database password.
   *
   * @public
   * @instance
   * @member {string} DB_PASSWORD
   */
  @IsString()
  @IsNotEmpty()
  public DB_PASSWORD: string

  /**
   * Database port.
   *
   * @default 27017
   *
   * @public
   * @instance
   * @member {number} DB_PORT
   */
  @IsNumber({ allowInfinity: false, allowNaN: false, maxDecimalPlaces: 0 })
  public DB_PORT: number

  /**
   * Database username.
   *
   * @default 'admin'
   *
   * @public
   * @instance
   * @member {string} DB_USERNAME
   */
  @IsString()
  @IsNotEmpty()
  public DB_USERNAME: string

  /**
   * HTTPS certfile content.
   *
   * @default ''
   *
   * @public
   * @instance
   * @member {string} HTTPS_CERT
   */
  @IsString()
  public HTTPS_CERT: string

  /**
   * HTTPS keyfile content.
   *
   * @default ''
   *
   * @public
   * @instance
   * @member {string} HTTPS_KEY
   */
  @IsString()
  public HTTPS_KEY: string

  /**
   * Node environment.
   *
   * @see {@linkcode NodeEnv}
   *
   * @default NodeEnv.DEV
   *
   * @public
   * @instance
   * @member {NodeEnv} NODE_ENV
   */
  @IsEnum(NodeEnv)
  public NODE_ENV: NodeEnv

  /**
   * Application URL (includes protocol).
   *
   * @default 'http://localhost'
   *
   * @public
   * @instance
   * @member {string} URL
   */
  @IsUrl({ require_protocol: true, require_tld: false })
  public URL: string

  /**
   * Creates a new configuration object instance.
   *
   * @param {NodeJS.ProcessEnv} [env={}] - Environment variables object
   */
  constructor({
    APP_ENV = AppEnv.DEV,
    DB_HOSTNAME = 'mongo',
    DB_NAME = APP_ENV,
    DB_PASSWORD = '',
    DB_PORT = '27017',
    DB_USERNAME = 'admin',
    HTTPS_CERT = '',
    HTTPS_KEY = '',
    NODE_ENV = NodeEnv.DEV,
    URL = 'http://localhost'
  }: NodeJS.ProcessEnv = {}) {
    this.APP_ENV = APP_ENV as AppEnv
    this.DB_HOSTNAME = DB_HOSTNAME
    this.DB_NAME = DB_NAME
    this.DB_PASSWORD = DB_PASSWORD
    this.DB_PORT = +DB_PORT
    this.DB_USERNAME = DB_USERNAME
    this.HTTPS_CERT = HTTPS_CERT
    this.HTTPS_KEY = HTTPS_KEY
    this.NODE_ENV = NODE_ENV as NodeEnv
    this.URL = URL

    // remove undefined properties
    Object.assign(this, shake(this, isUndefined))
  }

  /**
   * Config validator.
   *
   * @public
   *
   * @return {this} Validated configuration object
   * @throws {Exception<ValidationError>}
   */
  public validate(): Config {
    /**
     * Validation errors.
     *
     * @const {ValidationError[]} errors
     */
    const errors: ValidationError[] = validateSync(this, {
      enableDebugMessages: true,
      forbidUnknownValues: true,
      skipNullProperties: false,
      skipUndefinedProperties: false,
      stopAtFirstError: false,
      validationError: { target: false }
    })

    // throw if validation errors were encountered
    if (errors.length > 0) {
      throw new Exception<ValidationError>('Invalid configuration', {
        code: 'invalid-config',
        errors
      })
    }

    return this
  }
}

export default Config
