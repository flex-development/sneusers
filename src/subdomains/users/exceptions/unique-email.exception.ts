import { ObjectPlain } from '@flex-development/tutils'
import { ExceptionCode } from '@sneusers/enums'
import { Exception } from '@sneusers/exceptions'
import { IUserRaw } from '@sneusers/subdomains/users/interfaces'
import { UniqueConstraintError } from 'sequelize'

/**
 * @file Users Subdomain Exceptions - EmailException
 * @module sneusers/subdomains/exceptions/EmailException
 */

class UniqueEmailException extends Exception<UniqueConstraintError> {
  /**
   * Instantiates a new `UniqueEmailException`.
   *
   * @param {string} email - Conflicting user email address
   * @param {UniqueConstraintError} error - {@link UniqueConstraintError} thrown
   * @param {ObjectPlain} [data={}] - Custom error data
   */
  constructor(
    email: IUserRaw['email'],
    error: UniqueConstraintError,
    data: ObjectPlain = {}
  ) {
    super(ExceptionCode.CONFLICT)

    if (!data.message) {
      data.message = `User with email [${email}] already exists`
    }

    Object.assign(this, Exception.fromSequelizeError(error, data))
    Object.assign(this, { errors: error.errors })
  }
}

export default UniqueEmailException
