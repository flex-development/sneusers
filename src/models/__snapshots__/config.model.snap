// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:models/Config > constructor > should throw if config is invalid ({ HOST: " " }) 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isNotEmpty": "HOST should not be empty",
    },
    "property": "HOST",
    "value": "",
  },
]
`;

exports[`unit:models/Config > constructor > should throw if config is invalid ({ HOSTNAME: " " }) 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isNotEmpty": "HOSTNAME should not be empty",
    },
    "property": "HOSTNAME",
    "value": "",
  },
]
`;

exports[`unit:models/Config > constructor > should throw if config is invalid ({ NODE_ENV: "dev" }) 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isIn": "NODE_ENV must be one of the following values: development, test, production",
    },
    "property": "NODE_ENV",
    "value": "dev",
  },
]
`;
